e2e/nomostest/registry_server.go

```diff
@@ -0,0 +1,383 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nomostest
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest/gitproviders"
+	"kpt.dev/configsync/e2e/nomostest/registryproviders"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// RegistryUsername is a contrived username used for authenticating to the in-cluster registry
+	RegistryUsername = "user"
+	// RegistryPassword is a contrived password used for authenticating to the in-cluster registry
+	RegistryPassword = "password"
+	// RegistryHTTPPort is an HTTP port surfaced on the in-cluster registry.
+	RegistryHTTPPort = 5000
+	// RegistryHTTPSPort is an HTTPS port surfaced on the in-cluster registry. The HTTPS
+	// certificate is signed using a contrived CA.
+	RegistryHTTPSPort = 5001
+	// RegistryHTTPSAuthPort is an HTTPS port surfaced on the in-cluster registry with
+	// basic authentication enabled. The HTTPS certificate is signed using a contrived CA.
+	RegistryHTTPSAuthPort = 5002
+	// TestRegistryServer is the name of the in-cluster registry Deployment and Service
+	TestRegistryServer = "test-registry-server"
+	// TestRegistryServerAuthenticated is the name of the in-cluster registry Service
+	// which requires basic auth. This endpoint can be used as a sync URL when writing
+	// tests to verify basic auth.
+	TestRegistryServerAuthenticated = "test-registry-server-auth"
+	// TestRegistryNamespace is the namespace used for the in-cluster registry
+	TestRegistryNamespace = "test-registry-system"
+	nginxConfigMapName    = "nginx-cm"
+	nginxConf             = "nginx.conf"
+)
+
+func setupRegistry(nt *NT) error {
+	if *e2e.RegistryProvider == e2e.Local {
```

Comment:
not really specific to this PR, but should we ping the remote registry to validate auth and fail fast in the case that it's not local?